### PR TITLE
Flush before FlushAndTeardown in docs + examples

### PR DESCRIPTION
**Pull request by n8ta-msft**
Document this unexpected api requirement

### CODE REVIEW

File: examples/cpp/SampleCpp/main.cpp [hunk 1]
```diff
@@ -387,6 +388,8 @@ int main()
     // normal	- 4 sec
     // low		- 8 sec
 
+    printf("LogManager::Flush\n");
```
Comment (lalitb) [inline examples/cpp/SampleCpp/main.cpp:391]: Is this `printf` intended?

File: lib/include/public/ILogManager.hpp [hunk 2]
```diff
@@ -49,6 +49,9 @@ namespace MAT_NS_BEGIN
     {
        public:
         /// <summary>
+        /// WARNING: If you are using offline storage you need to call Flush before FlushAndTearDown.
+        /// https://github.com/microsoft/cpp_client_telemetry/blob/main/docs/use-after-free.md#flushandteardown
+        ///
```
Comment (lalitb) [inline lib/include/public/ILogManager.hpp:54]: While you are here, can you change the below comment too as `FlushAndTearDown` doesn't necessarily flush to persistent storage? I think just "tear-down the telemetry logging system" should be enough.

File: examples/cpp/SampleCpp/main.cpp (hunk 1 above)
Comment (n8ta-msft) [reply examples/cpp/SampleCpp/main.cpp:391]: I was just matching what was there for `    printf("LogManager::FlushAndTeardown\n");` I can remove it if you'd like

File: lib/include/public/ILogManager.hpp (hunk 2 above)
Comment (n8ta-msft) [reply lib/include/public/ILogManager.hpp:54]: Do you want me to modify the Flush() comment or remove the "Flushes any pending telemetry events in memory to disk" part of this comment?

**Review by lalitb — APPROVED** [review]

**Review by ThomsonTan — APPROVED** [review]
:shipit:

File: examples/cpp/SampleCppMini/main.cpp [hunk 3]
```diff
@@ -88,6 +88,7 @@ void test_cpp_api(const char * token, int ticketType, const char *ticket)
     logger->LogEvent(detailed_event);
 
     // Shutdown
+    LogManager::Flush();
```
Comment (maxgolov) [inline examples/cpp/SampleCppMini/main.cpp:91]: There's really something broken in the SDK if it requires `Flush` before `FlushAndTeardown`. The intent of `FlushAndTeardown` to perform `Flush`, wait for (configurable) N seconds, then `Teardown`.

**Comment by maxgolov** [comment]
> Document this unexpected api requirement

It's a bug.

File: examples/cpp/SampleCppMini/main.cpp (hunk 3 above)
Comment (lalitb) [reply examples/cpp/SampleCppMini/main.cpp:91]: Yes, that was what I thought too. But I don't see that's happening in FlushAndTeardown by looking into the code. Also we let this get documented here - https://github.com/microsoft/cpp_client_telemetry/blob/main/docs/use-after-free.md#flushandteardown

``
If one wishes to use this backstop to quiesce the SDK, the application should call Flush before FlushAndTeardown. This looks redundant (doesn't FlushAndTeardown actually flush, you might ask), but despite the method name FlushAndTeardown, the call to Flush is required to ensure persistence before the PauseActivity quiesce.
``

**Comment by n8ta-msft** [comment]
> > Document this unexpected api requirement
> 
> It's not unexpected requirement, it looks like a bug to me that was probably introduced around October 2021+

Glad to hear it's a bug, feel free to close my PR. Let me know if you'd like me to file an issue.

File: examples/cpp/SampleCppMini/main.cpp (hunk 3 above)
Comment (maxgolov) [reply examples/cpp/SampleCppMini/main.cpp:91]: I think this document was actually contributed at the same time when things got broken.. 🤣 I have a theory that this got broken by the zombie-logger change. My team uses a slightly older version of SDK, and we do not see the issue.

**Comment by lalitb** [comment]
> Glad to hear it's a bug, feel free to close my PR. Let me know if you'd like me to file an issue.

Thanks @n8ta-msft. Closing this. Feel free to file  a bug for better tracking :)

**Pull request closed without merging**